### PR TITLE
Support lint tool names in rustc command line options

### LINKED ISSUE #86628: Rustc does not parse tool name for lints controlled by the command line
**Problem**

This came up during the discussion of https://github.com/rust-lang/cargo/issues/5034. Lints can be broken into rustc builtin lints, and lints supplied by a tool such as clippy. For clippy lints, they are prefixed by `clippy::`. The idea is that external lints are supposed to be ignored if the tool is not active. This works for lints controlled through attributes, such as `#![allow(clippy::foo)]`.

Unfortunately the same thing doesn't work when allowing the lint through the command line, doing `rustc -A clippy::foo`.

**Steps**

1. Run `rustc -A clippy::foo test.rs`

**Possible Solution(s)**

The problem is at:
https://github.com/rust-lang/rust/blob/4733f32dbbbce7da5b06208a78af650be2db3001/compiler/rustc_lint/src/context.rs#L345
There is a hard-coded `None` passed in as the tool name in `check_lint_name`.

### PR DESCRIPTION
**Pull request by eholk**
When rustc is running without a lint tool such as clippy enabled, options for lints such as `clippy::foo` are meant to be ignored. This was already working for those specified by attrs, such as `#![allow(clippy::foo)]`, but this did not work for command line arguments like `-A clippy::foo`. This PR fixes that issue.

Note that we discovered this issue while discussing https://github.com/rust-lang/cargo/issues/5034.

Fixes #86628.

### CODE REVIEW

**Comment by camsteffen** [comment]
Does this mean we no longer need `--` in `cargo clippy -- -W clippy::lint`?

**Comment by eholk** [comment]
> Does this mean we no longer need `--` in `cargo clippy -- -W clippy::lint`?

It shoudn't have any change there. This simply makes it so you can do `rust -W clippy::lint` without getting an error message.

File: compiler/rustc_lint/src/context.rs [hunk 1]
```diff
@@ -1018,3 +1019,10 @@ impl<'tcx> LayoutOf for LateContext<'tcx> {
         self.tcx.layout_of(self.param_env.and(ty))
     }
 }
+
+pub fn parse_lint_and_tool_name(lint_name: &str) -> (Option<Symbol>, &str) {
+    match lint_name.split_once("::") {
+        Some((tool_name, lint_name)) => (Some(Symbol::intern(tool_name)), lint_name),
+        None => (None, lint_name),
+    }
+}
```
Comment (petrochenkov) [inline compiler/rustc_lint/src/context.rs:1028]: The correct way to implement this function is to factor out this code

https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/levels.rs#L324-L348

from `LintLevelsBuilder::push`.

The current implementation doesn't check whether the tool is known or not.

**Comment by petrochenkov** [comment]
The `src/tools/rust-analyzer` submodule update looks unintentional and needs to be removed.

File: compiler/rustc_lint/src/context.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_lint/src/context.rs:1028]: It looks like we have two places where we check if the tool is known: the place you mentioned, and also in `check_lint_name` (https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/context.rs#L414-L441). It'd probably be good to combine them, but I think the current implementation does check if the tool is known.

**Comment by eholk** [comment]
Thanks for the comments, @petrochenkov. I removed the rust-analyzer submodule update.

File: compiler/rustc_lint/src/context.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_lint/src/context.rs:1028]: >but I think the current implementation does check if the tool is known.

Could you add two test cases for that then?
```rust
// compile-flags: -A unknown_tool::foo <-- Error

fn main() {}
```
```rust
// compile-flags: -A known_tool::foo <-- Ok

#![register_tool(known_tool)]

fn main() {}
```

File: src/test/ui/lint/command-line-lint-tool.rs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+// compile-flags: -A clippy::foo
+// build-pass
```
Comment (petrochenkov) [inline src/test/ui/lint/command-line-lint-tool.rs:2]:
```suggestion
// check-pass
```
(No need to test codegen and linking here.)

File: compiler/rustc_lint/src/context.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_lint/src/context.rs:1028]: Ahh, I see my confusion. Clippy gets implicitly registered (it looks like [here]()), so the behavior is not to ignore lints for tools that are not defined. Clippy is just special.

The first test case you suggested looks like the same one I added, but it's supposed to fail while I assumed it was supposed to pass.

I'll add the new test cases and update my code to work the way it should.

**Comment by eholk** [comment]
This is a little better now, although probably not quite ready to merge. I think there's some more cleanup still to do. Is it at least on the right path for solving the problem?

I'm also having an issue with my tests. Even after adding the `.stderr` file, I get the following failure. Any ideas?
```
unexpected errors (from JSON output): [
    Error {
        line_num: 1,
        kind: Some(
            Error,
        ),
        msg: "1:1: 1:1: unknown lint tool: `unknown_tool` [E0602]",
    },
    Error {
        line_num: 1,
        kind: Some(
            Error,
        ),
        msg: "1:1: 1:1: unknown lint tool: `unknown_tool` [E0602]",
    },
]
```

**Comment by petrochenkov** [comment]
>I'm also having an issue with my tests. Even after adding the `.stderr` file, I get the following failure. Any ideas?

The tests are missing `//~ ERROR` annotations.
Although, since the errors are not attached to spans you'll probably have to use ``// error-pattern: unknown lint tool: `unknown_tool` `` instead.

File: src/test/ui/lint/command-line-register-unknown-lint-tool.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// compile-flags: -A unknown_tool::foo
+// check-fail
```
Comment (petrochenkov) [inline src/test/ui/lint/command-line-register-unknown-lint-tool.rs:2]: This comment doesn't do anything, tests are check-fail by default.

**Comment by petrochenkov** [comment]
Tests need some adjustments, but otherwise LGTM.

File: src/test/ui/lint/command-line-register-unknown-lint-tool.rs (hunk 3 above)
Comment (eholk) [reply src/test/ui/lint/command-line-register-unknown-lint-tool.rs:2]: Thanks, the error-pattern suggestion worked!

**Comment by eholk** [comment]
Okay, this should be ready to go now.

@bors -S-waiting-on-author +S-waiting-on-review

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**